numpy/__init__.pyi

```diff
@@ -1,15 +1,57 @@
-# very simple, just enough to start running tests
-#
 import builtins
-from typing import Any, Mapping, List, Optional, Tuple, Union
+
+from typing import (
+    Any, Dict, Iterable, List, Optional, Mapping, Sequence, Sized,
+    SupportsInt, SupportsFloat, SupportsComplex, SupportsBytes, SupportsAbs,
+    Tuple, Union,
+)
+
+import sys
 
 from numpy.core._internal import _ctypes
 
 _Shape = Tuple[int, ...]
 
+# Anything that can be coerced to a shape tuple
+_ShapeLike = Union[int, Sequence[int]]
+
+_DtypeLikeNested = Any  # TODO: wait for support for recursive types
+
+# Anything that can be coerced into numpy.dtype.
+# Reference: https://docs.scipy.org/doc/numpy/reference/arrays.dtypes.html
+_DtypeLike = Union[
+    dtype,
+    # default data type (float64)
+    None,
+    # array-scalar types and generic types
+    type,  # TODO: enumerate these when we add type hints for numpy scalars
+    # TODO: add a protocol for anything with a dtype attribute
+    # character codes, type strings or comma-separated fields, e.g., 'float64'
+    str,
+    # (flexible_dtype, itemsize)
+    Tuple[_DtypeLikeNested, int],
+    # (fixed_dtype, shape)
+    Tuple[_DtypeLikeNested, _ShapeLike],
+    # [(field_name, field_dtype, field_shape), ...]
+    List[Union[
+        Tuple[Union[str, Tuple[str, str]], _DtypeLikeNested],
+        Tuple[Union[str, Tuple[str, str]], _DtypeLikeNested, _ShapeLike]]],
+    # {'names': ..., 'formats': ..., 'offsets': ..., 'titles': ...,
+    #  'itemsize': ...} or {'field1': ..., 'field2': ..., ...}
+    # TODO: use TypedDict when/if it's officially supported
+    Dict[str, Any],
```

Comment:
Is this really `Any`? I think there could be:

```python
Union[
    Dict[str, Sequence[Union[int, str, _DtypeLike]]]   # for {"names": ..., "formats": ...}
    Dict[str, Union[Tuple[str, int], Tuple[_DtypeLike, int]]] # for {"field": ....}
]
```

Reply:
I think that should be
```

    # for {"names": ..., "formats": ...}
    Dict[str, Union[
        Sequence[str],  #names
        Sequence[_DtypeLike], # formats
        Sequence[int], # offsets
        int,  # itemsize
    ]],
    Dict[str, Tuple[_DtypeLike, int]] # for {"field": ....}
```

Reply:
This is probably dangerous because dict is an invariant type (so, for example, `Dict[str, int]` wouldn't be compatible with this big union type). Can you use `Mapping` instead?

Reply:
We could use `Mapping`, but it would be inaccurate. NumPy does actually insist on `dict`:
```
In [13]: od = collections.OrderedDict(x=(float, 1), y=(int, 2))

In [14]: np.dtype(od)
Out[14]: dtype({'names':['x','y'], 'formats':['<f8','<i8'], 'offsets':[1,2], 'itemsize':10})

In [15]: np.dtype(ChainMap([od]))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-15-03baf5fab980> in <module>()
----> 1 np.dtype(ChainMap([od]))

TypeError: data type not understood
```

Reply:
We have the same problem with:

```python
   List[Union[
       Tuple[Union[str, Tuple[str, str]], _DtypeLikeNested],
       Tuple[Union[str, Tuple[str, str]], _DtypeLikeNested, _ShapeLike]]],
```

which maybe should be `Sequence` but NumPy insists on taking a `list` (or at least, it doesn't take a tuple which is my go-to example of a non-list sequence).

Reply:
I added the more specific types for Dict, which should at least minimize the invariance issue.

Is it possible to make a dict that is covariant in its argument type? e.g.,
```python
K = TypeVar('V', covariant=True)
V = TypeVar('V', covariant=True)
def f(x: Dict[K, V]): ...
```